index.js

```diff
@@ -88,25 +86,63 @@ class PluginXrpPaychan extends PluginBtp {
       .toFixed(6, BigNumber.ROUND_UP)
   }
 
+  _setPeerAddress (peerAddress) {
+    if (!this._peerAddress && peerAddress) {
+      this._log.debug(`setting peer address to`, peerAddress)
+      this._peerAddress = peerAddress
+
+      const keyPairSeed = util.hmac(this._secret, CHANNEL_KEYS + this._peerAddress)
+      this._keyPair = nacl.sign.keyPair.fromSeed(keyPairSeed)
+    }
+  }
+
   async _setIncomingChannel (newId) {
-    if (!this._incomingChannel) {
-      this._log.debug('validating incoming paychan')
-      const details = await this._api.getPaymentChannel(newId)
-      this._validateChannelDetails(details)
-
-      this._log.trace('checking that peer\'s scale matches')
-      await this._getPeerInfo()
-
-      // second check occurs to prevent race condition where two lookups happen at once
-      if (!this._incomingChannel) {
-        this._log.debug(`setting incoming channel to`, newId)
-        this._log.trace(`channel details are`, details)
-        this._incomingChannel = newId
-        this._incomingChannelDetails = details
-        this._store.set('incoming_channel', this._incomingChannel)
-        await this._watcher.watch(this._incomingChannel)
+    if (newId === this._incomingChannel) {
+      return
+    }
+
+    // We're changing the details so don't accept packets.
+    this._reloadingChannel = true
+
+    this._log.trace('validating new paychan. id=' + newId)
+    const details = await this._api.getPaymentChannel(newId)
+    this._validateChannelDetails(details)
+
+    this._log.trace('checking that peer\'s scale matches')
+    await this._getPeerInfo()
+
+    // re-check in case of race conditions
+    if (this._incomingChannel && newId !== this._incomingChannel) {
```

Comment:
is the idea here behind these race conditions that the _setIncomingChannel function changes the _incomingChannel instance variable every time it is run, so we need to continue to check it's state after any async call?

Reply:
exactly, yeah